packages/studio-ui/src/AppWithLazyLoading.tsx

```diff
@@ -0,0 +1,75 @@
+import LoadingOverlay from "./components/LoadingOverlay";
+import { Suspense, lazy, useEffect, useState } from "react";
+import useStudioStore from "./store/useStudioStore";
+import ProgressBar from "./components/ProgressBar";
+import loadComponents from "./utils/loadComponents";
+import classNames from "classnames";
+
+const AppPromise = import("./App");
+const App = lazy(() => AppPromise);
+
+export default function AppWithLazyLoading() {
+  const [loadedCount, totalCount] = useStudioStore((store) => [
+    Object.keys(store.fileMetadatas.UUIDToImportedComponent).length,
+    Object.keys(store.fileMetadatas.UUIDToFileMetadata).length,
+  ]);
+  const componentsLoaded = loadedCount === totalCount;
+  const [appLoaded, setAppLoaded] = useState(false);
+
+  useEffect(() => {
+    loadComponents();
+    void AppPromise.then(() => setAppLoaded(true));
+  }, []);
+
+  return (
+    <LoadingOverlay
+      loading={!componentsLoaded || !appLoaded}
+      overlay={
+        <>
+          {renderComponentLoadingProgress(
+            loadedCount,
+            totalCount,
+            componentsLoaded
+          )}
+          {renderBundleMessage(appLoaded)}
+        </>
+      }
+    >
+      <Suspense>
+        <App />
+      </Suspense>
+    </LoadingOverlay>
+  );
+}
+
+function renderComponentLoadingProgress(
+  loadedCount: number,
+  totalCount: number,
+  componentsLoaded: boolean
+) {
+  const msg = componentsLoaded
+    ? "components loaded!"
+    : `loading components... (${loadedCount}/${totalCount})`;
+
+  return (
+    <>
+      <ProgressBar progressFraction={loadedCount / totalCount} />
+      <div
+        className={classNames("text-indigo-800", {
+          "animate-pulse": !componentsLoaded,
+        })}
+      >
+        {msg}
+      </div>
+    </>
+  );
+}
+
+function renderBundleMessage(appLoaded: boolean) {
+  const className = classNames("text-sky-600 mt-4", {
+    "animate-pulse": !appLoaded,
+  });
+  const msg = appLoaded ? "JS bundle loaded!" : "... loading JS bundle ...";
```

Comment:
Small nit: Can we come up with a more friendly message for non-technical users? They won't know what a JS bundle is.

Reply:
updated!